In [1]:
import pandas as pd
import numpy as np
import pickle

In [2]:
df = pd.read_csv('kaggle_diabetes.csv')

In [3]:
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            2      138             62             35        0  33.6   
1            0       84             82             31      125  38.2   
2            0      145              0              0        0  44.2   
3            0      135             68             42      250  42.3   
4            1      139             62             41      480  40.7   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.127   47        1  
1                     0.233   23        0  
2                     0.630   31        1  
3                     0.365   24        1  
4                     0.536   21        0

## Data Preprocessing

In [4]:
#remove zero value
totalzero = df == 0
print("Total Zero Values Are",totalzero.sum().sum())

Total Zero Values Are 3277


In [5]:
df_copy = df.copy(deep=True)
#convert zero to nan
df_copy[['Glucose','BloodPressure','SkinThickness','Insulin','BMI']] = df_copy[['Glucose','BloodPressure','SkinThickness','Insulin','BMI']].replace(0,np.NaN)

In [6]:
totalzerodf = df_copy == 0
print("After Convert Zero values",totalzerodf.sum().sum())

After Convert Zero values 1617


In [7]:
df_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               2000 non-null   int64  
 1   Glucose                   1987 non-null   float64
 2   BloodPressure             1910 non-null   float64
 3   SkinThickness             1427 non-null   float64
 4   Insulin                   1044 non-null   float64
 5   BMI                       1972 non-null   float64
 6   DiabetesPedigreeFunction  2000 non-null   float64
 7   Age                       2000 non-null   int64  
 8   Outcome                   2000 non-null   int64  
dtypes: float64(6), int64(3)
memory usage: 140.8 KB


In [8]:
#total null value
df_copy.isnull().sum()

Pregnancies                   0
Glucose                      13
BloodPressure                90
SkinThickness               573
Insulin                     956
BMI                          28
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64

In [9]:
#treat with null values
df_copy['Glucose'] = df_copy['Glucose'].fillna(df_copy['Glucose'].mean())
df_copy['BloodPressure'] = df_copy['BloodPressure'].fillna(df_copy['BloodPressure'].mean())
df_copy['SkinThickness'] = df_copy['SkinThickness'].fillna(df_copy['SkinThickness'].median())
df_copy['Insulin']=df_copy['Insulin'].fillna(df_copy['Insulin'].median())
df_copy['BMI'] = df_copy['BMI'].fillna(df_copy['BMI'].median())

In [10]:
df_copy.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [11]:
#Outier
# import seaborn as sns
# import matplotlib.pyplot as plt

# plt.subplot(511)
# sns.boxplot(df_copy['Glucose'])

# plt.subplot(512)
# sns.boxplot(df_copy['BloodPressure'])

# plt.subplot(513)
# sns.boxplot(df_copy['SkinThickness'])

# plt.subplot(514)
# sns.boxplot(df_copy['Insulin'])

# plt.subplot(515)
# sns.boxplot(df_copy['BMI'])

In [12]:
#split  data
X = df_copy.drop(columns='Outcome')
y = df_copy['Outcome']

In [13]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.20,random_state = 41)

## Model Build Logitic Regression

In [14]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr.fit(X_train,y_train)

C:\Users\Dell\AppData\Roaming\Python\Python37\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression()

In [29]:
y_pred = lr.predict(X_test)

In [30]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test,y_pred)
cm

array([[245,  21],
       [ 58,  76]], dtype=int64)

In [31]:
lr.score(X_test,y_test)

0.8025

### Logistic Regression score is 0.825

## Random Forest

In [18]:
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier()
rfc.fit(X_train,y_train)

RandomForestClassifier()

In [32]:
y_pred_rfc = rfc.predict(X_test)

In [20]:
cm_rfc = confusion_matrix(y_test,y_pred_rfc)
cm_rfc

array([[266,   0],
       [  2, 132]], dtype=int64)

In [21]:
rfc.score(X_test,y_test)

0.995

### Random Forest Score is 0.995

## So in between Logistic and Random forest i am use Random Forest Classifier for this problem

In [22]:
df_copy.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            2    138.0      62.000000           35.0    126.0  33.6   
1            0     84.0      82.000000           31.0    125.0  38.2   
2            0    145.0      72.403665           29.0    126.0  44.2   
3            0    135.0      68.000000           42.0    250.0  42.3   
4            1    139.0      62.000000           41.0    480.0  40.7   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.127   47        1  
1                     0.233   23        0  
2                     0.630   31        1  
3                     0.365   24        1  
4                     0.536   21        0

In [28]:
#result for Logisic Regression
result = lr.predict([[0, #Pregnancies
                  84.0, #Glucose
                  82.000000, #BloodPressure
                  31.0, #SkinThickness 
                  125.0, #Insulin
                  38.2, #BMI
                  0.233, #DiabetesPedigreeFunction
                  23 #Age
                     ]])
result
#logistic regression not give accurate

array([0], dtype=int64)

In [33]:
#result for Random Forest Classifier
result = rfc.predict([[0, #Pregnancies
                  84.0, #Glucose
                  82.000000, #BloodPressure
                  31.0, #SkinThickness 
                  125.0, #Insulin
                  38.2, #BMI
                  0.233, #DiabetesPedigreeFunction
                  23 #Age
                      ]])
result
#rfc give accurate

array([0], dtype=int64)

In [34]:
# Creating a pickle file for the classifier use randomforest classifier
filename = 'diabetes-prediction-rfc-model.pkl'
pickle.dump(rfc, open(filename, 'wb'))